crates/typst/src/layout/gridlayouter.rs

```diff
@@ -0,0 +1,812 @@
+    rrows: Vec<Vec<RowPiece>>,
+    /// Rows in the current region.
+    lrows: Vec<Row>,
+    /// The initial size of the current region before we started subtracting.
+    initial: Size,
+    /// Frames for finished regions.
+    finished: Vec<Frame>,
+    /// The span of the grid element.
+    span: Span,
+}
+
+/// The resulting sizes of columns and rows in a grid.
+#[derive(Debug)]
+pub struct GridLayout {
+    /// The fragment.
+    pub fragment: Fragment,
+    /// The column widths.
+    pub cols: Vec<Abs>,
+    /// The heights of the resulting rows segments, by region.
+    pub rows: Vec<Vec<RowPiece>>,
+}
+
+/// Details about a resulting row piece.
+#[derive(Debug)]
+pub struct RowPiece {
+    /// The height of the segment.
+    pub height: Abs,
+    /// The index of the row.
+    pub y: usize,
+}
+
+/// Produced by initial row layout, auto and relative rows are already finished,
+/// fractional rows not yet.
+enum Row {
+    /// Finished row frame of auto or relative row with y index.
+    Frame(Frame, usize),
+    /// Fractional row with y index.
+    Fr(Fr, usize),
+}
+
+impl<'a, T: Cell> GridLayouter<'a, T> {
+    /// Create a new grid layouter.
+    ///
+    /// This prepares grid layout by unifying content and gutter tracks.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        grid: &'a CellGrid<T>,
+        stroke: &'a Option<FixedStroke>,
+        regions: Regions<'a>,
+        styles: StyleChain<'a>,
+        span: Span,
+    ) -> Self {
+        // We use these regions for auto row measurement. Since at that moment,
+        // columns are already sized, we can enable horizontal expansion.
+        let mut regions = regions;
+        regions.expand = Axes::new(true, false);
+
+        Self {
+            grid,
+            has_gutter: grid.has_gutter,
+            stroke,
+            regions,
+            styles,
+            rcols: vec![Abs::zero(); grid.cols.len()],
+            width: Abs::zero(),
+            rrows: vec![],
+            lrows: vec![],
+            initial: regions.size,
+            finished: vec![],
+            span,
+        }
+    }
+
+    /// Determines the columns sizes and then layouts the grid row-by-row.
+    pub fn layout(mut self, engine: &mut Engine) -> SourceResult<GridLayout> {
+        self.measure_columns(engine)?;
+
+        for y in 0..self.grid.rows.len() {
+            // Skip to next region if current one is full, but only for content
+            // rows, not for gutter rows.
+            if self.regions.is_full() && (!self.has_gutter || y % 2 == 0) {
+                self.finish_region(engine)?;
+            }
+
+            match self.grid.rows[y] {
+                Sizing::Auto => self.layout_auto_row(engine, y)?,
+                Sizing::Rel(v) => self.layout_relative_row(engine, v, y)?,
+                Sizing::Fr(v) => self.lrows.push(Row::Fr(v, y)),
+            }
+        }
+
+        self.finish_region(engine)?;
+
+        self.render_fills_strokes()?;
```

Comment:
This will be called (and the for loop in `render_fills_strokes` will run) for all users of GridLayouter, including `list` and `enum`, which don't use fills and strokes at all. Before, I avoided that by checking if the global fill was `None`, but now I can't do that due to per-cell overrides.
Should I perhaps add some toggle in this `layout` method to opt into rendering fills and strokes, as to optimize for those use cases?

Reply:
I wouldn't worry about it for now. The overhead is probably relatively low.